fix propagation of channelRegistered event in ChannelInitializer [#4994]

tbcs opened the pull request:
This is a proposed fix for #4994 (please see the issue for further details).

tbcs commented:
There was a bug in the test case that I needed to fix (c56b09e).

normanmaurer commented:
@tbcs can you squash your commits ?

tbcs commented:
@normanmaurer Sure, I can squash the commits.  Is it OK to do a forced push to the topic branch?

normanmaurer commented:
@tbcs yes it is :)

normanmaurer commented:
@tbcs also please adjust the commit message to match our template:

http://netty.io/wiki/writing-a-commit-message.html

tbcs commented:
@normanmaurer Squashed, commit message adjusted, pushed.

normanmaurer commented on transport/src/main/java/io/netty/channel/ChannelInitializer.java:
```diff
@@ -65,9 +64,9 @@
     @SuppressWarnings("unchecked")
     public final void channelRegistered(ChannelHandlerContext ctx) throws Exception {
-        initChannel((C) ctx.channel());
         ctx.pipeline().remove(this);
```
actually I think the remove should still happen after the `initChannel(...)`

normanmaurer commented on transport/src/test/java/io/netty/channel/ChannelInitializerTest.java:
```diff
@@ -0,0 +1,105 @@
+        });
+        server.bind().sync();
+        client.connect(SERVER_ADDRESS).sync();
+        assertTrue(testSubject.channelRegisteredFired);
```
ensure you close the server and client channel.

normanmaurer commented on transport/src/test/java/io/netty/channel/ChannelInitializerTest.java:
```diff
@@ -0,0 +1,105 @@
+        });
+        server.bind().sync();
+        client.connect(SERVER_ADDRESS).sync();
+        assertTrue(testSubject.channelRegisteredFired);
```
ensure you close the server and client channel.

normanmaurer commented on transport/src/test/java/io/netty/channel/ChannelInitializerTest.java:
```diff
@@ -0,0 +1,105 @@
+        server.bind().sync();
+        client.connect(SERVER_ADDRESS).sync();
+        assertTrue(testSubject.channelRegisteredFired);
+    }
```
Also you can share most of the test-code and just use a different ChannelInitializer for each of them.

normanmaurer commented:
@tbcs please address my comments and ping me once done.

normanmaurer commented:
@Scottmitch @trustin @nmittler WDYT about always let the registration event start on the head of the pipeline after initChannel ? I think it makes perfect sense.

nmittler commented:
@normanmaurer wouldn't this result in handlers receiving the registered event multiple times?

normanmaurer commented:
@nmittler you mean if `ChannelInitializer` is not the only handler in the pipeline and other handlers exists before it? I think yes this may be possible, I wonder what would be a good way to fix this .

nmittler commented:
@normanmaurer not sure ... maybe we just accept that it may happen in strange corner cases?

normanmaurer commented:
@nmittler let me have a cold :beer: and think about ;)

nmittler commented:
@normanmaurer ha! The source of all good ideas! :)

Scottmitch commented:
I agree the possibility of multiple calls to channelRegistered will also create confusion, although it may be a less common use case. If `ChannelInitializer` doesn't want to support this use case perhaps we should explicitly check if `ChannelInitializer` is the first in the pipeline when `ChannelHandlerContext` is called. By default this could throw some sort of ISE saying the pipeline configuration is not supported by `ChannelInitializer` (with a comment about how channelRegistered may be called multiple times if allowed), and we `ChannelInitializer` could support an "opt out" mode to allow this behavior. @normanmaurer @nmittler - WDYT?

normanmaurer commented:
@trustin WDYT ? I'm not sure about this yet.

tbcs commented:
@normanmaurer I addressed the issues you raised earlier.  I also changed the test handler to count the registeredChannel events instead of just setting a flag, it is probably better to be explicit here.

normanmaurer commented:
@Scottmitch throwing an exception would be a API breakage... I think I have an idea... stand by :)

normanmaurer commented:
@trustin btw what is your stance on this ?

normanmaurer commented:
@trustin ping... at the moment I think we should just use `ChannelPipeline.fireChannelRegistered()` and log a warning if the `ChannelInitializer` is not the first handler in the pipeline. Alternative we may not log anything at all. WDYT ?

normanmaurer commented:
@Scottmitch @nmittler I'm leaning to just pull this in. I think this will cause the least surprising behaviour at least. WDYT ?

nmittler commented:
@normanmaurer it's probably fine ... @Scottmitch and @trustin, thoughts?

Scottmitch commented:
@normanmaurer @nmittler - no strong objection here

trustin commented:
Fine by me.

normanmaurer commented:
Cherry-picked into 4.1 (9d4fae308c09a2c759dc67db765d311154c8d977) and 4.0 (4fdf0e8026eab296c24ab41a89f2be16d7b428d9)

@tbcs thanks again!
